services/oauth/v2/oauth.go

```diff
@@ -0,0 +1,292 @@
+	if fetchErr != nil {
+		// TODO: should we have "nil" ?
+		return statusCode, nil, fetchErr
+	}
+	if statusCode == http.StatusOK {
+		// Token is available
+		// form refTokenParams & call RefreshToken()
+		fetchTokenParams.Secret = fetchSecret.Account.Secret
+	}
+	oauthHandler.RefreshToken(fetchTokenParams)
+	return oauthHandler.GetTokenInfo(fetchTokenParams, "Fetch token", authStats)
+}
+
+/*
+Refresh token function is used to refresh the token from the control plane
+- It fetches the token from the cache
+- If the token is present in the cache, it checks if the token has expired
+- If the token has expired, it fetches the token from the control plane
+- If the token is present in the cache, it compares the token with the received token
+  - If it doesn't match then it is possible the stored token is not expired but the received token is expired so return the stored token
+  - Else if matches then go ahead
+
+- It fetches the token from the control plane
+- It returns the status code, token and error
+- It also sends the stats to the statsd
+- It also sends the error to the control plane
+Parameters:
+  - refTokenParams: *RefreshTokenParams
+
+Return:
+  - int: status code
+  - *AuthResponse: token
+  - error: error
+*/
+func (authErrHandler *OAuthHandler) RefreshToken(refTokenParams *RefreshTokenParams) (int, *AuthResponse, error) {
+	authStats := &OAuthStats{
+		id:              refTokenParams.AccountId,
+		workspaceId:     refTokenParams.WorkspaceId,
+		rudderCategory:  "destination",
+		statName:        "",
+		isCallToCpApi:   false,
+		authErrCategory: REFRESH_TOKEN,
+		errorMessage:    "",
+		destDefName:     refTokenParams.DestDefName,
+		flowType:        authErrHandler.RudderFlowType,
+		action:          "refresh_token",
+	}
+	statusCode, refSecret, refErr := authErrHandler.GetTokenInfo(refTokenParams, "Refresh token", authStats)
+	// handling of refresh token response
+	if statusCode == http.StatusOK {
+		// refresh token successful --> retry the event
+		statusCode = http.StatusInternalServerError
+	} else {
+		// invalid_grant -> 4xx
+		// refresh token failed -> erreneous status-code
+		if refSecret.Err == REF_TOKEN_INVALID_GRANT {
+			// Add some kind of debug logger (if needed)
+			statusCode = authErrHandler.updateAuthStatusToInactive(refTokenParams.Destination, refTokenParams.WorkspaceId, refTokenParams.AccountId)
+		}
+	}
+	if refErr != nil {
+		refErr = fmt.Errorf("failed to refresh token: %w", refErr)
+	}
+	return statusCode, refSecret, refErr
+}
+
+func (oauthHandler *OAuthHandler) GetTokenInfo(refTokenParams *RefreshTokenParams, logTypeName string, authStats *OAuthStats) (int, *AuthResponse, error) {
+	startTime := time.Now()
+	defer func() {
+		authStats.statName = getOAuthActionStatName("total_latency")
+		authStats.isCallToCpApi = false
+		authStats.SendTimerStats(startTime)
+	}()
+	oauthHandler.Lock.Lock(refTokenParams.AccountId)
+	defer oauthHandler.Lock.Unlock(refTokenParams.AccountId)
```

Comment:
@chandumlg inside this getTokenIfo function the lock can be acquired by only one worker.